text/3373-avoid-nonlocal-definitions-in-fns.md

```diff
@@ -0,0 +1,63 @@
+- Feature Name: N/A
+- Start Date: 2022-01-19
+- RFC PR: [rust-lang/rfcs#3373](https://github.com/rust-lang/rfcs/pull/3373)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Starting in Rust 2024, stop allowing items inside functions to implement
+methods or traits that are visible outside the function.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, tools cross-referencing uses and definitions (such as IDEs) must
+either search inside all function bodies to find potential definitions
+corresponding to uses within a function, or not cross-reference those
+definitions at all.
+
+Humans cross-referencing such uses and definitions may find themselves
+similarly baffled.
+
+With this change, both humans and tools can limit the scope of their search and
+avoid looking for definitions inside other functions, without missing any
+relevant definitions.
+
+# Explanation
+[explanation]: #explanation
+
+Starting in the Rust 2024 edition:
+- An item nested inside a function or closure (through any level of nesting)
+  may not define an `impl Type` block unless the `Type` is also nested inside
+  the same function or closure.
+- An item nested inside a function or closure (through any level of nesting)
+  may not define an `impl Trait for Type` unless either the `Trait` or the
```

Comment:
What about `impl Inner<Outer> for Type` or `impl Trait for Inner<Outer>`?

Reply:
If `Inner` and `Outer` are non-local traits/types, these are just as problematic.

If `Inner` is a local (declared next to the impl) type or trait and `Outer` is non-local, I think it's fine to allow them because they won't be visible to the code outside.

Reply:
I meant "Inner" == "local", and "Outer" == "non-local".

Reply:
If only parameters of the trait/type are non-local I think it should be fine? Those impls still can't be observed from the outside since you'd need to be able to mention the trait/type still if I am not mistaken

Reply:
Not necessarily, you can actually leak local types to the outside like this: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1921433dd4209e89c7c96949a23e99b8

Reply:
Neither the trait nor the type in that impl are local to the function though, so that should be rejected by these restrictions.

Reply:
And potentially also cases like in https://internals.rust-lang.org/t/overly-strict-coherence-for-constrained-blanket-implementations/18204/8. I think there's like a whole coherence aspect to this?

Reply:
[You can still leak non-opaque types while following the proposed rules.](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a0e55cf93c0f4f12759604d9deb9c258)  If that seems far fetched, consider [`-> impl IntoIterator<Item = X>`](https://grep.app/search?q=-%3E%20impl%20IntoIterator) where the `IntoIter` associated type is not specified (but you're going to generate one).